SCMOD-3735: Add psql client and database creation script

michael-bryson opened the pull request:
https://jira.autonomy.com/browse/SCMOD-3735

Commit:
Adding psql client to base image and renamed to opensuse-base

rob-duewer reviewed (approved):
Looks fine to me.  Only question is why the version jump to 2.0.0? This is really just a process question, but I thought major version jumps were to be reserved for breaking changes?  Addition of psql client to the image won't break anything, so I would have guessed we were going to 1.2.0.  Is it because of the name change?

And last question.  Do we need to have a step to set all the script objects in /startup/startup.d/*.sh to executable?  I didn't check to see if the startup wrapper is handling this already, I'm just working from memory looking at some of the other derivative containers and wonder if it would be a useful safety to include.

michael-bryson commented:
@rob-duewer Yes the version jump is to do with the name change.

Re the scripts, once they are ready we will need to make a further change to the dockerfile to add them etc. I just created this change to get the client in there ready for when the scripts are added.

rob-duewer commented:
@michael-bryson Alrighty, all sounds good to me then.

dermot-hardy commented:
@michael-bryson I initially thought the name change was a breaking change also but perhaps we should just put this back to 1.1.0 to reflect that it is just a name change and there is no actual breaking change.

@rob-duewer There already is a step to set scripts executable --> https://github.com/CAFapi/opensuse-tini-image/blob/v1.0.2/src/main/docker/Dockerfile#L28

rob-duewer committed:
Added database creation script (#2)

Commit:
Moved version back to 1.1.0-SNAPSHOT and updated release notes

michael-bryson commented:
Changed version back to 1.1.0-SNAPSHOT and updated release notes.

dermot-hardy commented:
Looking at the output I think we need a change here as the current output makes it look like there is an error:

    startup.sh: Running startup scripts...
    startup.sh: Running check-create-pgdb.sh...
    check-create-pgdb.sh: Missing DATABASE
    check-create-pgdb.sh: Missing DATABASE_HOST
    check-create-pgdb.sh: Missing DATABASE_PORT
    check-create-pgdb.sh: Missing DATABASE_USERNAME
    check-create-pgdb.sh: Missing DATABASE_PASSWORD
    check-create-pgdb.sh: Not all required variables defined, exiting.
    check-create-pgdb.sh: HINT: If the ENV_PREFIX variable is provided, expected database parameters will be constructed with it.
    startup.sh: Running install-ca-cert-base.sh...
    install-ca-cert-base.sh: Not installing CA Certificate.
    startup.sh: Startup scripts completed

Commit:
Added CAF_ prefix to environment variables and renamed DATABASE to CAF_DATABASE_NAME

Commit:
SCMOD-3735: Only output missing variable messages if some are defined but not all

rob-duewer reviewed (dismissed):
I disagree with the most recent changes, will put details in comment shortly.

rob-duewer commented:
Details on rejection:
I think it is really important to list in the log the variables that are being considered and include in the log the hint about being able to specify a prefix.  Hiding this information unless all are not defined exclusively (and the bit about the optional prefix use completely) is a step backwards in terms of functionality and informing the administrator about what is being performed.  5 lines of log showing this is hardly a problem for regular operation.

That being said, Dermot's comment about making look less ominous is well taken.   So instead of changing the original logic to suppress these when all 5 are not defined, just change the messages reported.

So instead of "Missing "$(echo $ENV_PREFIX"DATABASE_NAME"), change to something like:
"INFO: Optional variable "$(echo $ENV_PREFIX"DATABASE_NAME") not defined."

Then when checking to make variables are present, have two types.  For case where all 5 are not defined, the new INFO text is fine if you also include the "HINT:" line.  (Prepend INFO: to it if you want.)  In this case, I assume you actually want to return 0 instead of 1, in case we later want to have startup.sh start paying attention to exit codes and stop when there's an error encountered.

Then for the case where some of them are not set, but some are; raise the warning level message (and include the HINT), exiting with a 1.

In addition, Dermot and I had previously discussed that we wanted NO default prefix.  This was purposely commented out and should remain so.  Given this is now much more broadly used by components other than just CAF_, (and the CAF name is fading as the team separation continues to be reduced) forcing them to be named this way for other consumers is not a good idea.  If an app wants to use the CAF_ prefix the base standard names; they can specify it explicitly like any other consumer.

And last thought.  The change to DATABASE_NAME as a base is fine with me, just that no other existing app that I surveyed was using that.  But I do think it makes more sense, and since if anyone wants to use the functionality, they're going to need to change for the others anyhow; I'm good with that part.

rob-duewer commented:
Ooh, one other thing I just thought of.  If we're going to use prepending the INFO: or WARNING: string; you probably want to set that in the other log areas too.  Like on log lines 65 and 71 (WARNING: and INFO: respectively), 129,133,137 (INFO, WARNING, INFO respectively), 154,156 (INFO,WARNING respectively).

And last comment (I promise).  Is WARN the usual standard instead of WARNING?  Not that it matters here of course, but I just figured we could be consistent with other tech if possible.

Commit:
SCMOD-3735: Addressing code review comments and added documentation

michael-bryson commented:
@rob-duewer Thanks for the feedback, I will try and address all your points:
- I have added the logs lines back in for each individual environment variable that is not defined and changed message as suggested.
- When none of the variables are defined I have added the HINT back in and changed the exit code to 0
- I have not added the HINT into the case where only some are defined because if the ENV_PREFIX is set to something different and only some are not found surely it is because they are not there and not that the prefix is different.
- Dermot and I discussed the default prefix yesterday and he was of the mind to have the prefix defaulting to CAF_ @dermot-hardy can you comment on this?
- I have changed WARNING to WARN as suggested and added INFO and WARN to the other echo statements you referred to on lines 65 and 71 as appropriate.

Commit:
Fixed table format

dermot-hardy commented:
I apologize if I agreed to no default prefix and am backtracking now but on reflection I feel that there should be a prefix because it's the only way to avoid the environment variable being too generic, and hence potentially acted on by some 3rd party libraries that we use, either now or in the future.  I must admit I don't particularly like 'CAF_DATABASE_XXX', not because I have any feeling that the CAF name is fading but rather just because it sounds like there is a thing called a CAF Database, which of course there isn't.  Let's chew on this one a while longer before we decide.

rob-duewer commented:
Looks better to me (although I don't see the new INFO/WARN added to the log lines in the check_db_exist and create_db functions.  Had you planned to bring those to snuff while you're there?).  I'll remove my decline vote.  :)

As for the prefix discussion, I would actually prefer to have no default prefix on this.  Dermot's point of possible confusion is well taken, especially if we were reacting to a single value or if we were trying to operate if we had some but not all expected vars defined.  But in this script, we check for all 5 of the needed variables before doing anything.  Since with each variable required check beyond one, the likleyhood of accidentally picking up one that isn't intentionally set by us decreases; I'm not concerned about it.  (Also that we're now back to always logging about the exact var names we were looking for and that we're reporting in both cases that some/none are defined mitigates this even further.)

Oh and finally, since this is a public base image that in theory people completely unrelated to us in CAF could use it as a base for their own non-Darwin image makes me think very much the general and descriptive names of DATABASE_xxx are perfect for this.  And in our case, where we do want to make them less general, having the optional ENV_PREFIX model meets both our and the general public's possible use cases well.  (At least the general public use cases I could think of quickly, of course. :))

Commit:
SCMOD-3735: Added informational prefixes to echo statements

michael-bryson commented:
Added informational prefixes to the echo statements in the other functions. The variable prefix conversation is still open.

dermot-hardy commented on README.md:
```diff
@@ -1,12 +1,31 @@
 ### Startup Scripts
 Any executable scripts added to the `/startup/startup.d/` directory will be automatically run each time the container is started (assuming the image entrypoint is not overwritten).
 
+#### Database Check and Create Script
```
It doesn't make sense to have this as a subheading under 'Startup Scripts' but for the Certificate Installation to be a level up.

Perhaps we should have a new 'Pre-Installed Startup Scripts' section and describe both scripts under it.

dermot-hardy commented on README.md:
```diff
@@ -1,12 +1,31 @@
 [Tini](https://github.com/krallin/tini) is pre-installed in the container.  If the image entrypoint is not overwritten then it will be automatically used.
 
+### PostgreSQL Client
+[PostgreSQL](https://wiki.postgresql.org/wiki/Main_Page) Client is pre-installed in the container. This enables projects to automatically install required PostgreSQL databases if required.
```
Link to https://www.postgresql.org/ or to the actual client page (i.e. https://www.postgresql.org/docs/current/static/app-psql.html) rather than to the wiki.

Instead of saying "This enables projects to automatically install required PostgreSQL databases if required.", use the description from the client page (i.e. "psql is a terminal-based front-end to PostgreSQL. It enables you to type in queries interactively, issue them to PostgreSQL, and see the query results. Alternatively, input can be from a file or from command line arguments. In addition, psql provides a number of meta-commands and various shell-like features to facilitate writing scripts and automating a wide variety of tasks.")

dermot-hardy commented on README.md:
```diff
@@ -1,12 +1,31 @@
 Any executable scripts added to the `/startup/startup.d/` directory will be automatically run each time the container is started (assuming the image entrypoint is not overwritten).
 
+#### Database Check and Create Script
+The `check-create-pgdb.sh` script is provided in the `/startup/startup.d/` directory by default. This script enables consumers of this image to define a PostgreSQL database that will be created during startup of the consuming container. The script will also check if the database already exists and if so will not attempt to create it.
```
"is provided in the `/startup/startup.d/` directory by default" --> "comes pre-installed"

dermot-hardy commented:
@rob-duewer Thinking about this whole prefix debate I think the root of the problem is that we are trying to make this a startup script.  How do you feel if we move this script into some sort of utilities or scripts folder in the container so that it is available to any service that wants to use it.  That way they could just call it in their service entrypoint script and pass in the desired prefix as an argument.  What do you think?

rob-duewer commented:
Possible I guess, but since this is safe to run all the time and we're including psql in the base; really think we want it run like the cert injection ones so the container owners don't have to do anything except define environment vars in their app definition.  Putting it in their specific entrypoint is exactly why we've got the cool flexible startup dropin model at the base image level in the first place.

Commit:
Updated readme as per Dermots comments

michael-bryson commented:
Addressed Dermots review comments in the readme document.

dermot-hardy reviewed (changes requested):
We need another `_APPNAME` variable for setting the application name.  This is important for establishing the source of database connections.  See [SCMOD-3077](https://jira.autonomy.com/browse/SCMOD-3077) for more details.

rob-duewer commented:
Ooh, very good idea.  (Although that won't be used in this script since psql exe does not give you control of that on connection establishment, having a standard env param name to specify it sounds like a very good idea to me for all those JDBC users out there.)

dermot-hardy commented:
Does setting the `PGAPPNAME` environment variable not work?

rob-duewer commented:
It might be possible (like having the trick of how to shove in the PW moderately safely...).  I just meant that it isn't an explicit psql command option.  Might be nice to have, but I don't see it as super important to the script's op.  It only makes one connection on launch (and I believe already reports itself as "psql admin" or some such, so it's not like there would lots of connections hanging around for it to have to sort through.  But if you wanna try adding it, I'm not at all against it.  (Although the app had better be using this same variable to determine its connection or else it's pretty useless. :) )

Commit:
Merge branch develop into install_psql_client

michael-bryson commented:
@dermot-hardy @rob-duewer We need a decision on this as this story is in the current Gamma sprint which finishes tomorrow. We need agreement on environment variable prefixes and also the _APPNAME or PGAPPNAME variable so I can get this finished off before the end of the sprint.

dermot-hardy commented:
Sorry Michael - I'll take a look at it later on today.

Commit:
Merge branch 'develop' into install_psql_client

Commit:
Removed CAF_ references from DB script

Commit:
Moved database creation script into a separate folder

Commit:
Updated to pass the prefix as a mandatory argument

Commit:
Updated readme to reflect changes in variable prefix

dermot-hardy commented:
I've moved the script out of the startup scripts folder and made it available for services to use from their entrypoint scripts.  When the script is called the prefix to use needs to be passed in as an argument.

@michael-bryson Please update the documentation to reflect this change in usage.

Commit:
SCMOD-3735: Updated documentation to reflect new functionality

michael-bryson commented:
@dermot-hardy I have updated the documentation, can you review this please?

Commit:
Escape the arguments when checking them

Commit:
Added another variable for specifying the application name

dermot-hardy commented:
@michael-bryson Please do some testing around the escaping of the arguments.  For example, can we create a database with a quote in the name?

michael-bryson commented:
@dermot-hardy Single quotes and double quotes do not work in the database name and throw errors in the SQL commands. We need to ensure that we are protecting against potential sql injection here.

rob-duewer commented:
+1 On checking for validity on incoming variables, especially that var name existence check is now quoted.  A simple regex to make sure there are no funny chars coming in the vars seems prudent.  Database name of course, but also database user and the rest (perhaps all but database password.)  I would actually reuse these proposed checking functions into the other scripts that consume environment vars as well.  (in caf-common/container-cert-script/ for example).

+0 On including the application name for the PSQL connection.  I really think this is pretty irrelevant for this task.  If it is defined for the actual application to use for it's database connections, then I'm fine with the script using it.  But requiring it to be provided for this script to function is overkill.

-1 Requiring a prefix I am strongly against.  As this script does not inject any env variables to the container, it has no bearing on the actual application that might be using them to mean something else.  And if that app is using them for something else, this script is intended to be safe to run every time.  (If something is provided that is invalid the connection to a SQL target, it will catch and display those errors and do no harm.)

-1 Moving it out of the autorun facility I also strongly disagree with.  The whole purpose of that facility and bundling safe-to-always-run-utilities is to allow app creators not to have to make any additions to their entry point definitions.  This script, like the cert scripts, falls into this category.

So I am voting to reject those modifications for the above reasons.  But as you own this project; you make the call to override these objections or not.  (With your modifications, the data-management-service can still use the resulting image. So that is the base case we need.  However, with them, i suspect very few other projects based on this image will make use of this new useful feature.  Which I think is quite unfortunate.)

rob-duewer reviewed (changes requested):
See comment above.

Commit:
Corrected the log level for of a message

Commit:
Updated to escape the database name properly

Commit:
Renamed internal variables to match environment variables

Commit:
Added DATABASE_APPNAME to the dummy section

Commit:
Further update to the readme

Commit:
Updated release notes
